internal/mesh/internal/controllers/sidecarproxy/controller.go

```diff
@@ -230,6 +238,95 @@ func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req c
 	return nil
 }
 
+func (r *reconciler) workloadPortProtocolsFromService(
+	ctx context.Context,
+	fetcher *fetcher.Fetcher,
+	workload *types.DecodedWorkload,
+) (map[string]*pbcatalog.WorkloadPort, error) {
+
+	// Fetch all services for this workload.
+	serviceIDs := r.cache.ServicesForWorkload(workload.GetResource().GetId())
+
+	services := make([]*types.DecodedService, len(serviceIDs))
+
+	for i, serviceID := range serviceIDs {
+		svc, err := fetcher.FetchService(ctx, serviceID)
+		if err != nil {
+			return nil, err
+		}
+
+		// If service is not found, we should untrack it.
+		if svc == nil {
+			r.cache.UntrackService(serviceID)
+			continue
+		}
+
+		services[i] = svc
+	}
+
+	// Now walk through all workload ports.
+	// For ports that don't have a protocol explicitly specified, inherit it from the service.
+
+	result := make(map[string]*pbcatalog.WorkloadPort)
+
+	for portName, port := range workload.GetData().GetPorts() {
+		if port.GetProtocol() != pbcatalog.Protocol_PROTOCOL_UNSPECIFIED {
+			// Add any specified protocols as is.
+			result[portName] = port
+			continue
+		}
+
+		if len(serviceIDs) == 0 || len(services) == 0 {
+			result[portName] = &pbcatalog.WorkloadPort{
+				Port:     port.GetPort(),
+				Protocol: pbcatalog.Protocol_PROTOCOL_TCP,
+			}
+			continue
+		}
+
+		// Otherwise, look for port protocol in the service.
+		inheritedProtocol := pbcatalog.Protocol_PROTOCOL_UNSPECIFIED
+		for _, svc := range services {
+			// Find workload's port as the target port.
+			svcPort := svc.GetData().FindServicePort(portName)
+
+			// If this service doesn't select this port, skip.
+			if svcPort == nil {
+				continue
+			}
+
+			// Check for conflicts.
+			// If protocols between services selecting this workload on this port do not match,
+			// we need to report this mismatch to the user and use the default protocol (tcp) instead.
```

Comment:
Do you need to bubble up a condition to report?

Reply:
I thought about it but it's too hard so I punted on it for now. The other reason is that it's not super clear where to report it and whether k8s users would even see it. If I report it on a workload, k8s users won't be able to see it because we can't report workload statuses back to pods. Maybe proxy state template would make more sense 🤷‍♀️ 

I could add some logging for now though just so we have at least some way to debug this.